### PR TITLE
Update gacha mechanism

### PR DESCRIPTION
**Pull request by ghost**
## Description

- A new probability formula is used, where N is soft pity and M is hard pity.  
P(t) = P_base when t <= N  
P(t) = (1 + P_base)*(t - N)/(M - N) + P_base when t > N  

- Add soft pity of purple items,default is 9.

- Add 4-star featured character pity, if the first 4-star item you win in this event wish is not one of the featured characters or weapons, then the next 4-star item you win is guaranteed to be a featured character or weapons.

- Add Thoma to standard banner.

## Type of changes

- [x] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by 1-lavender-1** [comment]
"P(t) = 1 + P_base*(1 - (t - N)/(M - N)) when t > N"
it seems like P(t)>100% when t>N?

**Comment by ghost** [comment]
> “P（t） = 1 + P_base*（1 - （t - N）/（M - N）） 当 t > N” 当 t（t）>100% 当 t>N？
Sorry, it should be  ( 1 - P_base ) * ( t - N）/（M - N）+ P_base

File: src/main/java/emu/grasscutter/game/gacha/GachaBanner.java [hunk 1]
```diff
@@ -18,9 +18,11 @@ public class GachaBanner {
 	private int[] rateUpItems2;
 	private int baseYellowWeight = 60; // Max 10000
 	private int basePurpleWeight = 510; // Max 10000
-	private int eventChance = 50; // Chance to win a featured event item
-	private int softPity = 75;
+	private int eventChance = 5000; // Chance to win a featured event item
+	private int softPity = 74; // Increase the rate at this
 	private int hardPity = 90;
+	private int prupleSoftPity = 9;
```
Comment (ghost) [inline src/main/java/emu/grasscutter/game/gacha/GachaBanner.java]: You got a typo here.

**Review by ghost — CHANGES_REQUESTED** [review]
Thank you for wanting to contribute, but the gacha system rework is already being done by @Birdulon in #639 so im not sure if I will accept this PR unless you are adding something useful that he isnt.

File: src/main/java/emu/grasscutter/game/gacha/GachaBanner.java (hunk 1 above)
Comment (ghost) [reply src/main/java/emu/grasscutter/game/gacha/GachaBanner.java]: ok

**Comment by victor-lcr** [comment]
Can I use this function in the latest compilation ?

**Comment by ghost** [comment]
Thank you for wanting to contribute to our project, but i will be closing this PR as these changes were added in #639

**Pull request closed without merging**